### PR TITLE
CPP: Deal with two very similar Japanese era queries

### PR DESCRIPTION
**Pull request by geoffw0**
Similar to https://github.com/Semmle/ql/pull/1838, this PR deals with two user contributed queries that have an identical name.  Following the rename, I decided to combine the two queries into one since there is really no reason a user would want to use one query but not the other, or distinguish between their results.  The resulting query is also in a more logical existing directory.

### CODE REVIEW

File: cpp/ql/src/Likely Bugs/JapaneseEra/ConstructorOrMethodWithExactEraDate.ql [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
+ * @name Hard-coded Japanese era start date in call
  * @description Japanese era changes can lead to code behaving differently. Avoid hard-coding Japanese era start dates.
  * @kind problem
  * @problem.severity warning
  * @id cpp/japanese-era/constructor-or-method-with-exact-era-date
  * @precision medium
  * @tags reliability
  *       japanese-era
+ * @deprecated This query is deprecated, use
+ *             Best Practices/Magic Constants/JapaneseEraDate.ql instead.
```
Comment (jbj) [inline cpp/ql/src/Likely Bugs/JapaneseEra/ConstructorOrMethodWithExactEraDate.ql]: I think the modern way we refer to queries is by `@id`.

File: change-notes/1.23/analysis-cpp.md [hunk 2]
```diff
@@ -8,13 +8,15 @@ The following changes in version 1.23 affect C/C++ analysis in all applications.
 
 | **Query**                   | **Tags**  | **Purpose**                                                        |
 |-----------------------------|-----------|--------------------------------------------------------------------|
-| Query name (`query id`) | tags | Message. |
+| Hard-coded Japanese era start date (`cpp/japanese-era/exact-era-date`) | reliability, japanese-era | This query is a combination of two old queries that were identical in purpose but separate as an implementation detail.  This new query replaces Hard-coded Japanese era start date in call (`cpp/japanese-era/constructor-or-method-with-exact-era-date`) and Hard-coded Japanese era start date in struct (`cpp/japanese-era/struct-with-exact-era-date`). |
```
Comment (jbj) [inline change-notes/1.23/analysis-cpp.md]: We did not actually announce the two existing queries in the 1.22 change notes. See https://semmle.slack.com/archives/C7NS4H8HY/p1566982840001100.

@felicity-semmle, if you think this change note should be structured differently, let's discuss it on Slack.

File: cpp/ql/src/Likely Bugs/JapaneseEra/ConstructorOrMethodWithExactEraDate.ql (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/JapaneseEra/ConstructorOrMethodWithExactEraDate.ql]: Good point.  I will update this and the place I copied it from.

File: change-notes/1.23/analysis-cpp.md (hunk 2 above)
Comment (geoffw0) [reply change-notes/1.23/analysis-cpp.md]: Would you prefer to just have a change note about the 'new' query, or no change note at all for this change?

Comment (semmledocs-ac) [reply change-notes/1.23/analysis-cpp.md]: As those queries were provided by the community I think the current change note is appropriate. We don't want to fail to give credit to the original author by suggesting this is a brand new query.

Comment (jbj) [reply change-notes/1.23/analysis-cpp.md]: I'm fine with the change note as it is.

**Comment by semmledocs-ac** [comment]
qhelp LGTM

File: cpp/ql/src/Best Practices/Magic Constants/JapaneseEraDate.ql [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+predicate badStructInitialization(Element target, string message) {
+  exists(StructLikeClass s, YearFieldAccess year, MonthFieldAccess month, DayFieldAccess day |
+    assignedYear(s, year, 1989) and
+    assignedMonth(s, month, 1) and
+    assignedDay(s, day, 8) and
+    target = year and
+    message = "A time struct that is initialized with exact Japanese calendar era start date."
+  )
+}
+
```
Comment (zlaski-semmle) [inline cpp/ql/src/Best Practices/Magic Constants/JapaneseEraDate.ql]: In addition to 1989-08-01, should we also handle 2019-05-01?  See https://en.wikipedia.org/wiki/Japanese_era_name.

File: cpp/ql/src/Best Practices/Magic Constants/JapaneseEraDate.qhelp [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+  <p>
+    When eras change, date and time conversions that rely on a hard-coded era start date need to be reviewed. Conversions relying on Japanese dates in the current era can produce an ambiguous date. 
+    The values for the current Japanese era dates should be read from a source that will be updated, such as the Windows registry.
```
Comment (rdmarsh2) [inline cpp/ql/src/Best Practices/Magic Constants/JapaneseEraDate.qhelp]: We should add some guidance on handling this on Linux and MacOS systems as well.

File: cpp/ql/src/Best Practices/Magic Constants/JapaneseEraDate.qhelp [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  </p>
+</overview>
+
+<references>
+  <li>
+    <a href="https://blogs.msdn.microsoft.com/shawnste/2018/04/12/the-japanese-calendars-y2k-moment/">The Japanese Calendar's Y2K Moment</a>.
```
Comment (rdmarsh2) [inline cpp/ql/src/Best Practices/Magic Constants/JapaneseEraDate.qhelp]: And some Linux and MacOS references as well.

File: cpp/ql/src/Best Practices/Magic Constants/JapaneseEraDate.ql (hunk 3 above)
Comment (geoffw0) [reply cpp/ql/src/Best Practices/Magic Constants/JapaneseEraDate.ql]: Good idea.  I'll do this in a follow-up PR as it's a separate thing from cleaning up the duplication.

File: cpp/ql/src/Best Practices/Magic Constants/JapaneseEraDate.qhelp (hunk 4 above)
Comment (geoffw0) [reply cpp/ql/src/Best Practices/Magic Constants/JapaneseEraDate.qhelp]: Do you know what the right guidance might be?

File: cpp/ql/src/Likely Bugs/JapaneseEra/ConstructorOrMethodWithExactEraDate.ql (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/JapaneseEra/ConstructorOrMethodWithExactEraDate.ql]: Done.

File: cpp/ql/src/Best Practices/Magic Constants/JapaneseEraDate.ql (hunk 3 above)
Comment (geoffw0) [reply cpp/ql/src/Best Practices/Magic Constants/JapaneseEraDate.ql]: https://github.com/Semmle/ql/pull/1867

**Pull request merged**